README.md
# ap

THIS IS A WORK IN PROGRESS... USE AT YOUR OWN RISK!

    > cat test/known-scenario-1.txt
    activity act-01 5
    activity act-02 3
    activity act-03 5
    activity act-04 3
    activity act-05 3
    activity act-06 3
    activity act-07 3
    activity act-08 5
    activity act-09 3
    activity act-10 5
    activity act-11 5
    activity act-12 5 act-03
    activity act-13 10
    person user-01 80
    person user-02 20
    person user-03 60
    person user-04 80
    person user-05 60
    person user-06 20
    person user-07 80
    claim user-01 act-01
    claim user-02 act-02
    claim user-03 act-03
    claim user-04 act-04
    claim user-05 act-05
    claim user-06 act-06

    > cat test/known-scenario-1.txt | ap --today 2020-03-30
    act-04 2020-03-30 2020-04-02 user-04
    act-09 2020-03-30 2020-04-02 user-07
    act-05 2020-03-30 2020-04-06 user-05
    act-01 2020-03-30 2020-04-07 user-01
    act-03 2020-03-30 2020-04-09 user-03
    act-02 2020-03-30 2020-04-20 user-02
    act-06 2020-03-30 2020-04-20 user-06
    act-11 2020-04-02 2020-04-13 user-07
    act-13 2020-04-02 2020-04-21 user-04
    act-08 2020-04-06 2020-04-16 user-05
    act-10 2020-04-07 2020-04-15 user-01
    act-07 2020-04-09 2020-04-16 user-03
    act-12 2020-04-13 2020-04-21 user-07

    > cat test/known-scenario-1.txt | ap --today 2020-03-30 | ap-gantt
    # This should pop Gnuplot open!

## Input

### Activities

    activity act-01 10
    activity act-02 5
    activity act-03 5 act-01 act-02

Activity entries are prefixed with the 'activity' keyword, and they specify:

- the activity name
- the effort (in days) required to complete the activity
- an optional list activities that the current one depends on

### People

    person user-01 90
    person user-02 50

Person entries are prefixed with the 'person' keyword, and they specify:

- the person name
- their allocation to _this_ project, with the idea being that the smaller their
  allocation, the longer it's going to take for them to complete a given
  activity

### Claims

    claim user-03 act-01

Claim entries are prefixed with the 'claim' keyword, and they specify:

- the person who has claimed..
- the activity that was calimed

### Out of office

    out-of-office user-01 2020-04-25
    out-of-office user-01 2020-12-25

Out of office entries are prefixed with the 'out-of-office' keyword, and they
specify:

- the person this OOO entry refers to
- the date (YYYY-MM-DD) the person won't be up for working

# TODO

- Find a good heuristic
- when people start off with multiple activities assigned, `ap` combines them
  together -- which means other activities depending on any of them, might end
  up being claimed later because the completion date was pushed too far ahead
  (i.e. when the two activities got completed)
- Shared, and mutual resources
- Docker support?!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- First release
- `--today DATE` option to run the planning as if today was `DATE`
- `--productivity 100` option to configure how productive people will be (max
  is `100`)
- `--disable-skip-weekends` option to have the planner schedule work for
  weekends as well as business days
- `--good-enough` option to have the planner use some aggressive heuristics to
  speed up the computation (yes, it might return a sub-optimal schedule)


src/test.lisp
(format t "~a/~a~&" (lisp-implementation-type) (lisp-implementation-version))

;; By adding the current directory to ql:*local-project-directories*, we can
;; QL:QUICKLOAD this without asking users to symlink this repo inside
;; ~/quicklisp/local-projects, or clone it right there in the first place.
(push #P"." ql:*local-project-directories*)
(ql:quickload :ap :silent T)

(ql:quickload :ap/tests :verbose T)
(let ((exit-code 0))
  (handler-case (asdf:test-system :ap)
    (error (c)
      (format T "~&~A~%" c)
      (setf exit-code 1)))
  (uiop:quit exit-code))


test/main.lisp
(defpackage :ap/tests
  (:use :cl :pmdb :ap.quickutils :ap.utils :ap)
  (:export #:run))

(in-package :ap/tests)

(defvar *tests* nil "A list of tests; the default argument to `run'.")

(defun run () (1am:run *tests*))

(defmacro deftest (name &body body)
  "Define a test function and add it to `*tests*`."
  `(progn
     (let ((1am:*tests* nil))
       (1am:test ,name ,@body))
     (pushnew ',name *tests*)))

(deftest ap/simple/completion-day
  (1am:is (= 15 (completion-day (uiop:read-file-string #P"test/simple.txt")
                                :disable-skip-weekends t))))

(deftest ap/simple/completion-day/disable-find-optimal-solution
  (1am:is (= 15 (completion-day (uiop:read-file-string #P"test/simple.txt")
                                :disable-skip-weekends t
                                :disable-find-optimal-solution t))))

(deftest ap/simple/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/simple.txt")
                                :disable-skip-weekends t
                                :ignore-claims t
                                :disable-find-optimal-solution t))))


(deftest ap/dependencies/completion-day
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/dependencies.txt")
                                :disable-skip-weekends t))))

(deftest ap/dependencies/completion-day/disable-find-optimal-solution
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/dependencies.txt")
                                :disable-skip-weekends t
                                :disable-find-optimal-solution t))))

(deftest ap/dependencies/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/dependencies.txt")
                                :disable-skip-weekends t
                                :ignore-claims t
                                :disable-find-optimal-solution t))))


(deftest ap/multiple-claims-per-person/completion-day
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/multiple-claims-per-person.txt")
                                :disable-skip-weekends t))))

(deftest ap/multiple-claims-per-person/completion-day/disable-find-optimal-solution
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/multiple-claims-per-person.txt")
                                :disable-skip-weekends t
                                :disable-find-optimal-solution t))))

(deftest ap/multiple-claims-per-person/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 5 (completion-day (uiop:read-file-string #P"test/multiple-claims-per-person.txt")
                               :disable-skip-weekends t
                               :ignore-claims t
                               :disable-find-optimal-solution t))))


(deftest ap/out-of-office/completion-day
  (1am:is (= 8 (completion-day (uiop:read-file-string #P"test/out-of-office.txt")
                               :today "2020-04-28"))))

(deftest ap/out-of-office/completion-day/disable-find-optimal-solution
  (1am:is (= 8 (completion-day (uiop:read-file-string #P"test/out-of-office.txt")
                               :today "2020-04-28"
                               :disable-find-optimal-solution t))))

(deftest ap/out-of-office/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 8 (completion-day (uiop:read-file-string #P"test/out-of-office.txt")
                               :today "2020-04-28"
                               :ignore-claims t
                               :disable-find-optimal-solution t))))


(deftest ap/known-scenario-1/completion-day
  (1am:is (= 22.25 (completion-day (uiop:read-file-string #P"test/known-scenario-1.txt")
                                   :today "2020-03-30"))))

(deftest ap/known-scenario-1/completion-day/disable-find-optimal-solution
  (1am:is (= 22.25 (completion-day (uiop:read-file-string #P"test/known-scenario-1.txt")
                                   :today "2020-03-30"
                                   :disable-find-optimal-solution t))))

; (deftest ap/known-scenario-1/completion-day/ignore-claims/disable-find-optimal-solution
;   (1am:is (= 22.25 (completion-day (uiop:read-file-string #P"test/known-scenario-1.txt")
;                                    :today "2020-03-30"
;                                    :ignore-claims t
;                                    :disable-find-optimal-solution t))))


(deftest ap/known-scenario-2/completion-day
  (1am:is (= 21 (completion-day (uiop:read-file-string #P"test/known-scenario-2.txt")
                                :today "2020-04-10"))))

(deftest ap/known-scenario-2/completion-day/disable-find-optimal-solution
  (1am:is (= 21 (completion-day (uiop:read-file-string #P"test/known-scenario-2.txt")
                                :today "2020-04-10"
                                :disable-find-optimal-solution t))))

(deftest ap/known-scenario-2/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 19.75 (completion-day (uiop:read-file-string #P"test/known-scenario-2.txt")
                                   :today "2020-04-10"
                                   :ignore-claims t
                                   :disable-find-optimal-solution t))))


(deftest ap/known-scenario-3/completion-day
  (1am:is (= 11.5 (completion-day (uiop:read-file-string #P"test/known-scenario-3.txt")
                                  :today "2020-04-24"))))

(deftest ap/known-scenario-3/completion-day/disable-find-optimal-solution
  (1am:is (= 11.5 (completion-day (uiop:read-file-string #P"test/known-scenario-3.txt")
                                  :today "2020-04-24"
                                  :disable-find-optimal-solution t))))

(deftest ap/known-scenario-3/completion-day/ignore-claims/disable-find-optimal-solution
  (1am:is (= 11.5 (completion-day (uiop:read-file-string #P"test/known-scenario-3.txt")
                                  :today "2020-04-24"
                                  :ignore-claims t
                                  :disable-find-optimal-solution t))))


(deftest ap/known-scenario-4/completion-day
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/known-scenario-4.txt")
                                :today "2020-05-01"))))

(deftest ap/known-scenario-4/completion-day/disable-find-optimal-solution
  (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/known-scenario-4.txt")
                                :today "2020-05-01"
                                :disable-find-optimal-solution t))))

; (deftest ap/known-scenario-4/completion-day/ignore-claims/disable-find-optimal-solution
;   (1am:is (= 10 (completion-day (uiop:read-file-string #P"test/known-scenario-4.txt")
;                                 :today "2020-05-01"
;                                 :ignore-claims t
;                                 :disable-find-optimal-solution t))))


build.lisp
(format t "~a/~a~&" (lisp-implementation-type) (lisp-implementation-version))

(ql:quickload :deploy :silent T)

;; By adding the current directory to ql:*local-project-directories*, we can
;; QL:QUICKLOAD this without asking users to symlink this repo inside
;; ~/quicklisp/local-projects, or clone it right there in the first place.
(push #P"." ql:*local-project-directories*)
(ql:quickload :ap :silent T)

(setf ap:*version* (let* ((system (asdf:find-system :ap nil))
                          (base-version (asdf:component-version system))
                          (git-cmd (format NIL "git rev-list v~a..HEAD --count" base-version))
                          (output (uiop:run-program git-cmd
                                                    :output :string
                                                    :ignore-error-status T))
                          (pending (parse-integer output :junk-allowed T)))
                     (if (or (not pending) (zerop pending))
                       (format NIL "~a" base-version)
                       (format NIL "~a-r~a" base-version pending))))

(setf deploy:*status-output* nil)
(pushnew :deploy-console *features*)

(let ((deploy:*status-output* t))
  (asdf:make :ap :force t))


src/main.lisp
(in-package #:ap)

(defvar *version* nil "Application version")
(defvar *ignore-claims* nil "Ignore any person-activity claim")
(defvar *person-productivity* 1 "Person productivity")
(defvar *today* nil "First day of the simulation")
(defvar *skip-weekends* t "Skip weekends when calculating days of effort")
(defvar *find-optimal-solution* t "When disabled, quicker search algorithms will be run, potentially
  settling for sub-optimal solutions")

; Date utils --------------------------------------------------------------------------------------

(defun get-universal-time-start-of-day ()
  "Similar to GET-UNIVERSAL-TIME, except with seconds, minutes, hour and timezone slots,
  all set to 0"
  (destructuring-bind (second minute hour timezone) (list 0 0 0 0)
    (multiple-value-bind (sec min hr day month year)
        (decode-universal-time (get-universal-time))
      (declare (ignore sec min hr))
      (encode-universal-time second minute hour day month year timezone))))

(defun parse-date (str)
  "Parses a YYYY-MM-DD string, and returns its universal-time value.

  => (parse-date \"1900-01-01\")
  0

  => (parse-date \"1986-11-28\")
  2742508800"
  (destructuring-bind (second minute hour timezone) (list 0 0 0 0)
    (destructuring-bind (year month day)
        (mapcar #'parse-integer (split-sequence:split-sequence #\- str))
      (encode-universal-time second minute hour day month year timezone))))

(defun offset-to-universal-time (offset)
  "Adds OFFSET days to *TODAY*, and return its universal time representation.

  Note: if OFFSET is not a natural number, (FLOOR OFFSET) will be used instead."
  (+ *today* (* (floor offset) (* 24 60 60))))

(defun offset-is-weekend-p (offset)
  "Returns T if *TODAY* plus OFFSET days falls on a weekend"
  (multiple-value-bind (sec min hr day month year dow dst-p tz)
      (decode-universal-time (offset-to-universal-time offset))
    (declare (ignore sec min hr day month year dst-p tz))
    (>= dow 5)))

(defun offset-is-ooo-p (offset ooo-calendar &aux (time (offset-to-universal-time offset)))
  "Returns T if OOO-CALENDAR contains an entry for *TODAY* plus OFFSET days."
  (multiple-value-bind (value existsp)
      (gethash time ooo-calendar)
    (declare (ignore value))
    existsp))

(defun offset-add-business-days (offset calendar n)
  "Adds N business days to OFFSET.

  If OFFSET points to a Sunday, and N is 0, the function will return next Monday's offset.

  If N is negative, the function will ERROR out."
  (when (< n 0)
    (error "The number of business days to add, N, cannot be negative: ~a" n))
  (recursively ((offset offset)
                (n n))
    (cond ((and *skip-weekends* (offset-is-weekend-p offset) (recur (1+ offset) n)))
          ((offset-is-ooo-p offset calendar) (recur (1+ offset) n))
          (T (if (zerop n)
               offset
               (let ((inc (min n 1)))
                 (recur (+ offset inc) (- n inc))))))))

(defun offset-to-date (offset)
  "Converts an offset to a YYYY-MM-DD string."
  (multiple-value-bind (sec min hr day mon yr dow dst-p tz)
      (decode-universal-time (offset-to-universal-time offset))
    (declare (ignore sec min hr dow dst-p tz))
    (format nil "~d-~2,'0d-~2,'0d" yr mon day)))


; Bits utils --------------------------------------------------------------------------------------

(defun bit-set-p (j number)
  "Returns T if the j-th bit of NUMBER, is set."
  (= (logandc2 (ash 1 j) number) 0))

(defun set-bit (j number)
  "Sets j-th bit of NUMBER, and return the result."
  (logior number (ash 1 j)))

(defun all-bits-set (number)
  (loop
    :for j :from 0
    :for mask = (ash 1 j)
    :until (> mask number)
    :when (bit-set-p j number)
    :collect j))

; Input / parsing ---------------------------------------------------------------------------------

(defun read-from-stream (s)
  (loop
    :for line = (read-line s NIL :eof)
    :until (eq line :eof)
    :collecting line :into lines
    :finally (return (format nil "~{~a~^~&~}" lines))))

(defstruct activity id effort depends-on)

(defun parse-float (s)
  (with-input-from-string (in s)
    (* 1.0 (read in))))

(defun parse-percentage (s &aux (percentage (/ (parse-float s) 100)))
  (unless (and (<= 0 percentage 1))
    (error "~a not in the range [0, 100]" s))
  percentage)

(defun parse-activity (s &aux (parts (split-sequence:split-sequence #\Space s)))
  (make-activity :id (second parts)
                 :effort (parse-float (third parts))
                 :depends-on (nthcdr 3 parts)))

(defstruct person id allocation)

(defun parse-person (s &aux (parts (split-sequence:split-sequence #\Space s)))
  (make-person :id (second parts)
               :allocation (parse-percentage (third parts))))

(defstruct claim person activity)

(defun parse-claim (s &aux (parts (split-sequence:split-sequence #\Space s)))
  (make-claim :person (second parts)
              :activity (third parts)))

(defstruct ooo person date)

(defun parse-ooo (s &aux (parts (split-sequence:split-sequence #\Space s)))
  (make-ooo :person (second parts)
            :date (parse-date (third parts))))

(defstruct simulation
  activities
  dependencies
  people
  calendars
  already-working-on
  already-been-busy-for)

(defun days-to-complete (person activity)
  (with-slots (allocation) person
    (with-slots (effort) activity
      (/ effort allocation *person-productivity*))))

(defun parse-simulation (string)
  (let ((person-id-map (make-hash-table :test 'equal))
        (activity-id-map (make-hash-table :test 'equal))
        activities
        people
        claims
        ooo-entries)
    (loop
      :for line :in (mapcar #'trim (split-sequence:split-sequence #\Newline string))
      :until (zerop (length line))
      :when (string= (subseq line 0 8) "activity") :do (push (parse-activity line) activities)
      :when (string= (subseq line 0 6) "person") :do (push (parse-person line) people)
      :when (and (not *ignore-claims*)
                 (string= (subseq line 0 5) "claim")) :do (push (parse-claim line) claims)
      :when (string= (subseq line 0 13) "out-of-office") :do (push (parse-ooo line) ooo-entries))
    (setf activities (reverse activities)
          people (reverse people))
    (let ((calendars (map-into (make-array (length people)) #'make-hash-table))
          (dependencies (make-array (length activities) :initial-element 0))
          (already-working-on (make-array (length people) :initial-element 0))
          (already-been-busy-for (make-array (length people) :initial-element 0)))
      (loop
        :for i :below (length people)
        :for person :in people
        :do (setf (gethash (person-id person) person-id-map) i))
      (loop
        :for ooo :in ooo-entries
        :for person-id = (ooo-person ooo)
        :for i = (gethash person-id person-id-map)
        :unless i :do (error "Cannot find a person with name: ~a, when parsing ~a" person-id ooo)
        :do (let ((person-calendar (aref calendars i)))
              (setf (gethash (ooo-date ooo) person-calendar) T)))
      (loop
        :for i :below (length activities)
        :for activity :in activities
        :do (setf (gethash (activity-id activity) activity-id-map) i))
      (loop
        :for claim :in claims
        :for person-id = (claim-person claim)
        :for i = (gethash person-id person-id-map)
        :for activity-id = (claim-activity claim)
        :for j = (gethash activity-id activity-id-map)
        :unless i :do (error "Cannot find a person with name: ~a, when parsing ~a" person-id claim)
        :unless j :do (error "Cannot find an activity with name: ~a, when parsing ~a" activity-id claim)
        :do (setf (aref already-working-on i) (set-bit j (aref already-working-on i))
                  (aref already-been-busy-for i) (offset-add-business-days
                                                   (aref already-been-busy-for i)
                                                   (aref calendars i)
                                                   (days-to-complete (nth i people) (nth j activities)))))
      (loop
        :for i :below (length activities)
        :for activity :in activities
        :do (loop
              :for dep-id :in (activity-depends-on activity)
              :for j = (gethash dep-id activity-id-map)
              :unless j :do (error "Cannot find ~a's dependency: ~a" (activity-id activity) dep-id)
              :do (setf (aref dependencies i)
                        (set-bit j (aref dependencies i)))))
      (make-simulation :activities (coerce activities 'vector)
                       :dependencies dependencies
                       :people (coerce people 'vector)
                       :calendars calendars
                       :already-working-on already-working-on
                       :already-been-busy-for already-been-busy-for))))

(defun trim (string)
  (string-trim
    '(#\Space #\Newline #\Backspace #\Tab
      #\Linefeed #\Page #\Return #\Rubout)
    string))

; Search ------------------------------------------------------------------------------------------

(defstruct (cost-info (:conc-name nil)) days dependencies)

(defun precompute-costs (sim)
  (with-slots (activities dependencies people) sim
    (let* ((n (length people))
           (m (length activities))
           (costs (make-array (list n m))))
      (loop
        :for i :below n
        :for person :across people
        :do (loop
              :for j :below m
              :for activity :across activities
              :for days = (days-to-complete person activity)
              :do (setf (aref costs i j) (make-cost-info :days days
                                                         :dependencies (aref dependencies j)))))
      costs)))

(defstruct (worker (:conc-name nil)) been-working-on been-busy-for)
(defstruct (state (:conc-name nil)) workers completed complete-dates)

(defun change (seq i el)
  (let ((c (copy-seq seq)))
    (setf (aref c i) el)
    c))

(defun target-date (state)
  (maximization (workers state) :key #'been-busy-for))

(defun dependencies-completed-p (dependencies completed)
  (= (logandc2 dependencies completed) 0))

(defun dependencies-already-completed-p (dependencies completed complete-dates been-busy-for)
  (and (dependencies-completed-p dependencies completed)
       (loop
         :for j :in (all-bits-set dependencies)
         :for complete-date = (aref complete-dates j)
         :always (>= been-busy-for complete-date))))

(defun neighbors (costs calendars activity-count state)
  (let ((target-date (target-date state)))
    (with-slots (workers completed complete-dates) state
      (loop
        :for i :below (length workers)
        :for c :across calendars
        :for w :across workers
        :for been-busy-for = (been-busy-for w)
        :appending (loop
                     :for j :below activity-count
                     :for cost = (aref costs i j)
                     :when (and
                             cost
                             (not (bit-set-p j completed))
                             (dependencies-already-completed-p (dependencies cost) completed complete-dates been-busy-for))
                     :collecting (let* ((been-busy-for (offset-add-business-days been-busy-for c (days cost)))
                                        (next-worker (make-worker :been-working-on (set-bit j (been-working-on w))
                                                                  :been-busy-for been-busy-for))
                                        (workers (change workers i next-worker))
                                        (complete-dates (change complete-dates j been-busy-for))
                                        (next-state (make-state :workers workers
                                                                :completed (set-bit j completed)
                                                                :complete-dates complete-dates))
                                        (next-target-date (target-date next-state)))
                                   (cons
                                     next-state
                                     (max 0 (- next-target-date target-date)))))))))

(defun activities-remaining (activities state)
  (with-slots (completed) state
    (loop
      :for j :below (length activities)
      :for a :across activities
      :when (not (bit-set-p j completed))
      :collect a)))

(defun effort-remaining (activities state)
  (reduce #'+ (activities-remaining activities state)
          :key #'activity-effort))

(defun heuristic (activities state)
  "Works with standard tests // ignore claims and it won't help much
  Works with lots-of-work"
  (let* ((effort-remaining (effort-remaining activities state)))
    effort-remaining))

; Output ------------------------------------------------------------------------------------------

(defun dummy-initial-state (worker-count)
  (make-state :workers (make-array worker-count
                                   :initial-element (make-worker :been-working-on 0
                                                                 :been-busy-for 0))))

(defun create-shedule (sim steps &aux (end-state (nth (1- (length steps)) steps)))
  (let* ((worker-count (length (workers end-state))))
    (loop
      :for (s1 s2) :on (cons (dummy-initial-state worker-count) steps)
      :while s2
      :appending (loop
                   :for i :below worker-count
                   :for w1 :across (workers s1)
                   :for w2 :across (workers s2)
                   :for changed = (- (been-working-on w2) (been-working-on w1))
                   :unless (zerop changed)
                   :appending (loop
                                :for j :in (all-bits-set changed)
                                :collect (list
                                           (activity-id (aref (simulation-activities sim) j))
                                           (been-busy-for w1)
                                           (been-busy-for w2)
                                           (person-id (aref (simulation-people sim) i))))))))

(defun schedule-activities (sim-string)
  (let* ((sim (parse-simulation sim-string))
         (completed (reduce #'logior (simulation-already-working-on sim)))
         (workers (coerce (loop
                            :for been-working-on :across (simulation-already-working-on sim)
                            :for been-busy-for :across (simulation-already-been-busy-for sim)
                            :collect (make-worker :been-working-on been-working-on
                                                  :been-busy-for been-busy-for))
                          'vector))
         (calendars (simulation-calendars sim))
         (activity-count (length (simulation-activities sim)))
         (all-activities (1- (ash 1 activity-count)))
         (costs (precompute-costs sim))
         (complete-dates (loop
                           :with result = (make-array activity-count :initial-element -1)
                           :for been-working-on :across (simulation-already-working-on sim)
                           :for been-busy-for :across (simulation-already-been-busy-for sim)
                           :unless (zerop been-working-on)
                           :do (loop
                                 :for j :in (all-bits-set been-working-on)
                                 :do (setf (aref result j) been-busy-for))
                           :finally (return result)))
         (init-state (make-state :workers workers
                                 :completed completed
                                 :complete-dates complete-dates))
         end-state end-state-cost end-state-path)
    (setf (values end-state end-state-cost end-state-path)
          (if *find-optimal-solution*
            (a* init-state
                :init-cost (target-date init-state)
                :goalp (lambda (state) (= (completed state) all-activities))
                :neighbors (partial-1 #'neighbors costs calendars activity-count)
                :test 'equalp)
            (multiple-value-bind (best-state best-state-cost best-state-path)
                (a* init-state
                    :init-cost (target-date init-state)
                    :goalp (lambda (state) (= (completed state) all-activities))
                    :neighbors (partial-1 #'neighbors costs calendars activity-count)
                    :heuristic (partial-1 #'heuristic (simulation-activities sim))
                    :test 'equalp)
              (multiple-value-bind (end-state end-state-cost end-state-path)
                  (dfs init-state
                       :init-cost (target-date init-state)
                       :goalp (lambda (state) (= (completed state) all-activities))
                       :neighbors (partial-1 #'neighbors costs calendars activity-count))
                (if (< end-state-cost best-state-cost)
                  (values end-state end-state-cost end-state-path)
                  (values best-state best-state-cost best-state-path))))))
    (when end-state
      (values
        end-state
        (create-shedule sim end-state-path)))))

(defun sort-schedule(schedule)
  (sort
    (copy-seq schedule)
    (lambda (a1 a2)
      (if (< (second a1) (second a2))
        T
        (if (= (second a1) (second a2))
          (< (third a1) (third a2)))))))

(defun pprint-schedule (schedule &aux (schedule (sort-schedule schedule)))
  (loop
    :for (activity-id from to person-id) :in schedule
    :do (format t "~a ~a ~a ~a~&"
                activity-id
                (offset-to-date from)
                (offset-to-date to)
                person-id)))

; Options -----------------------------------------------------------------------------------------

(opts:define-opts
  (:name :help
   :description "print the help text and exit"
   :short #\h
   :long "help")
  (:name :version
   :description "print the version and exit"
   :short #\v
   :long "version")
  (:name :today
   :description "use DATE (YYYY-MM-DD) as first day of the simulation (defaults to today)"
   :long "today"
   :arg-parser #'parse-date
   :meta-var "DATE")
  (:name :productivity
   :description "percentage of day spent by people working on activities (defaults to 100)"
   :long "productivity"
   :arg-parser #'parse-percentage
   :meta-var "PROD")
  (:name :disable-skip-weekends
   :description "Don't skip weekends when calculating actual days of efforts."
   :long "disable-skip-weekends")
  (:name :good-enough
   :description "Trade execution time, for a potentially sub-optimal solution."
   :long "good-enough"))

(define-condition exit (error)
  ((code
     :initarg :code
     :initform 0
     :reader exit-code))
  (:report (lambda (condition stream)
             (format stream "Trying to exit with code: ~S"
                     (exit-code condition)))))

(defun parse-opts (argv)
  (multiple-value-bind (options)
      (handler-case
          (handler-bind ((opts:missing-required-option (lambda (condition)
                                                         (if (or (member "-h" argv :test #'equal)
                                                                 (member "--help" argv :test #'equal)
                                                                 (member "-v" argv :test #'equal)
                                                                 (member "--version" argv :test #'equal))
                                                           (invoke-restart 'opts:skip-option)
                                                           (progn
                                                             (format t "~a~%" condition)
                                                             (error 'exit :code 1)))))
                         (opts:arg-parser-failed (lambda (condition)
                                                   (format t "~a~%" condition)
                                                   (error 'exit :code 1))))
            (opts:get-opts argv))
        (opts:unknown-option (condition)
          (format t "~a~%" condition)
          (error 'exit :code 1))
        (opts:missing-arg (condition)
          (format t "~a~%" condition)
          (error 'exit :code 1)))
    (if (getf options :help)
      (progn
        (opts:describe
          :prefix "XXX"
          :args "[keywords]")
        (error 'exit)))
    (if (getf options :version)
      (progn
        (format T "~a~%" *version*)
        (error 'exit)))
    ; optional ones
    (if (getf options :disable-skip-weekends)
      (setf *skip-weekends* nil))
    (setf *today* (or (getf options :today) (get-universal-time)))
    (setf *person-productivity* (or (getf options :productivity) 1))
    (setf *find-optimal-solution* (not (getf options :good-enough)))))

; API ---------------------------------------------------------------------------------------------

(defun completion-day (sim-string &key today ignore-claims
                                  disable-skip-weekends
                                  disable-find-optimal-solution)
  (let ((*ignore-claims* ignore-claims)
        (*skip-weekends* (not disable-skip-weekends))
        (*today* (or (and today (parse-date today)) (get-universal-time)))
        (*find-optimal-solution* (not disable-find-optimal-solution)))
    (multiple-value-bind (end-state)
        (schedule-activities sim-string)
      (target-date end-state))))

(defun toplevel ()
  (handler-case (parse-opts (opts:argv))
    (exit (condition)
      (opts:exit (exit-code condition))))
  (multiple-value-bind (end-state schedule)
      (handler-case (schedule-activities (read-from-stream *standard-input*))
        (error (condition)
          (format t "Unexpected error: ~a~&" condition)
          (opts:exit -1)))
    (declare (ignore end-state))
    (pprint-schedule schedule)))
